Feature/python wrapping

imikejackson opened the pull request.

Commit:
Fixes to Python wrapping so that sample pipelines could be run entirely within Python

Commit:
Made the data in DataArrays accessible in Python

Commit:
Python helper function to convert numpy array into DataArray

Commit:
Added CreateGeometry helper function for Python

Commit:
Added CMake commands to copy Python-related files and dependencies to directory specified with PYTHON_SITE_PACKAGES entry

Commit:
Added new CMake cache variables for site packages and dependencies directories. Updated dependencies to be correct for build configuration type.

Commit:
Added CreateAsciiWizardData and CreateDynamicTableData helper functions for Python

Commit:
Fixed an issue with the data type for input( table data) in CreateDynamicTableData helper function.

Commit:
Added RemoveArray(s) and MultiThresholdObjects helper functions for Python

Commit:
Added MultiThresholdObjects2 and MoveData helper functions for Python

Commit:
Added Python bindings and helper functions to support generation of StatsData in Synthetic Building pipelines.

Commit:
Fix so the copying of the SIMPL Python module code and dependencies goes to a subfolder, dream3d

Commit:
Added enumerated types for several of the filters to the SIMPLib module, if possible, or as a convenience enum in Python (simpl_common.py).
Fixed an issue with the use of the Python keyword "lambda" in some filters by renaming the method arguments for the Pythonic functions.

Commit:
Improvements and fixes to enumeration types for Python binding

Commit:
Added helper function for creation of Data Container Proxies. Split the custom type casters into a separate header file.

imikejackson committed:
Python: Updating modules to have a more sane naming scheme.

Needs testing within a Python IDE like Pycharm

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

Commit:
Added some of the example pipelines in Python

imikejackson committed:
PYTHON: Use the binary directory as the 'site-packages' directory.

This will fix the issue of making copies of all dependent libraries being copied
on Windows to allow runtime to work.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
PYTHON: Use the output directories as the python folder.

Update the PyBind11Generator to write to the proper location.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Set the archive output name for the Python module.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
PYTHON: More refinements to the python interface.

The naming scheme has been shortened where possible.
Ghosting of modules has been fixed

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

Commit:
Updates to the example pipelines, addition of enumeration for delimiter types, and improvements to the ConvertToDataArray helper functions in Python

Commit:
Fixed the import statement for simpl_helpers Python file as it was based on previous module naming scheme

Commit:
Adding additional pipeline examples in Python and updated existing ones to PEP8 formatting standards.

Commit:
Added Python bindings for EnsembleInfo, FileListInfo, Crystal Structure enums and QJsonArrays
Added most of the remaining example pipelines that require ITK Image Processing to be installed.

Commit:
Added additional Python binding for PrecipitateStatsData, DatasetImportInfo list, NumericTypes and AngleUnits. Added new example pipelines to test filters that are not used in any current example pipelines.

Commit:
Added additional Python examples for some Core and DREAM3DReview filters. Fixed the helper function for DeleteData to accept just Data Containers or Attribute Matrices. Added an enum for ObjectToCopy

Commit:
PYTHON: Added CMake code to build a Conda package in Windows. The package can be uploaded with an appropriate .condarc file in the user's home directory to the Ancaonda package repository.

Commit:
PYTHON: Added an option to install the Conda package locally to the same Conda environment used to build it.

Commit:
PYTHON: Removed unnecessary output from a custom type cast. Added code to remove existing dream3d packages before installing a new one (When SIMPL_PYTHON_INSTALL_CONDA_PACKAGE_LOCALLY is true)

Commit:
PYTHON: Fixing mistake in previous commit

Commit:
PYTHON: Removed unnecessary output for DatasetImportInfoList Python binding. Fixed an issue with local package installation.

Commit:
PYTHON: Changes to Conda package building to improve support for Linux/Mac

Commit:
PYTHON: Added a Python script to test all the example pipelines in the Examples folder

imikejackson committed:
Fix issue where clang-format messed up the spacing in the macro.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fix deprecation warnings and clang-tidy warnings

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
separate windows specific paths from other operating systems

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Make the DREAM3D/SIMPLView/SIMPL CMake consistent with features.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/AttributeMatrixProxy.cpp:
```diff
@@ -285,3 +285,64 @@ void AttributeMatrixProxy::updatePath(DataArrayPath::RenameType renamePath)
+// -----------------------------------------------------------------------------
+DataArrayProxy& AttributeMatrixProxy::getDataArrayProxy(const QString& name)
+{
+	for (QMap<QString, DataArrayProxy>::iterator iter = dataArrays.begin(); iter != dataArrays.end(); ++iter) // DataArray Level
```
Check your indentation for this entire section.  Either you are using tabs or a very different tab-to-spaces ratio than everywhere else.

mmarineBlueQuartz replied:
There's also no newline at the end of the file, which we have everywhere else.  I'm not sure if we care about consistency for that, but apparently GitHub makes note of it.

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/AttributeMatrixProxy.h:
```diff
@@ -137,6 +149,58 @@ class SIMPLib_EXPORT AttributeMatrixProxy
+	/**
+	 * @brief Get the attribute matrices (Python Binding)
+	 */
+	QMap<QString, DataArrayProxy> getdataArrays();
```
This chunk of new code is indented an additional time.

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/DataArrayProxy.cpp:
```diff
@@ -324,3 +324,13 @@ void DataArrayProxy::updatePath(DataArrayPath::RenameType renamePath)
+
+void DataArrayProxy::setflag(uint8_t newFlag)
+{
+	flag = newFlag;
```
Both flag methods need an indentation check, and there's no new line at the end (again, not sure if we care about that)

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/DataContainerArrayProxy.cpp:
```diff
@@ -544,3 +544,14 @@ QMap<QString, DataContainerProxy> DataContainerArrayProxy::readMap(QJsonArray js
+
+ void DataContainerArrayProxy::setdataContainers(QMap<QString, DataContainerProxy> newDataContainers)
+{
+	 dataContainers = newDataContainers;
```
Indentation for goth dataContainer methods. 
Check capitalization on `getdataContainers` and `setdataContainers`.  If that is intentional, it should be marked in both .h and .cpp because it does not look like it is and invites anyone working on the file to "fix" it.

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/DataContainerArrayProxy.h:
```diff
@@ -121,6 +126,17 @@ class SIMPLib_EXPORT DataContainerArrayProxy
+	 * @brief Get the data containers (Python Binding)
+	 * @param new dataContainers map
+	 */
+	 void setdataContainers(QMap<QString, DataContainerProxy>);
```
Indentation

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/DataContainerProxy.cpp:
```diff
@@ -235,3 +235,64 @@ void DataContainerProxy::setFlags(uint8_t flag, AttributeMatrixProxy::AMTypeFlag
+
+QMap<QString, AttributeMatrixProxy> DataContainerProxy::getattributeMatricies()
+{
+	return attributeMatricies;
```
This entire chunk needs to be checked for indentation.
Are these methods supposed to be lowercase until the third word, because I still want to "fix" that.  I'm starting to think I should not.

mmarineBlueQuartz replied:
Not sure if we care, but there's no new line at the end of the file.

mmarineBlueQuartz replied:
getAttributeMatrices through getAttributeMatrixProxy still need indentation checks.  Also, none of them have the three-line comment blocks before their definitions.

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/DataContainerProxy.h:
```diff
@@ -127,6 +140,52 @@ class SIMPLib_EXPORT DataContainerProxy
+	/**
+	 * @brief Get the attribute matrices (Python Binding)
+	 */
+	QMap<QString, AttributeMatrixProxy> getattributeMatricies();
```
Indentation

mmarineBlueQuartz commented on Source/SIMPLib/Filtering/AbstractComparison.h:
```diff
@@ -47,6 +47,8 @@
 class SIMPLib_EXPORT AbstractComparison
 {
+	PYB11_CREATE_BINDINGS(AbstractComparison)
+	PYB11_PROPERTY(int UnionOperator READ getUnionOperator WRITE setUnionOperator)
```
Indentation

mmarineBlueQuartz commented on Source/SIMPLib/StatsData/PrecipitateStatsData.h:
```diff
@@ -83,6 +83,21 @@ typedef QVector<FloatArrayType::Pointer> VectorOfFloatArray;
+	PYB11_PROPERTY(float PhaseFraction READ getPhaseFraction WRITE setPhaseFraction)
+	PYB11_PROPERTY(float BoundaryArea READ getBoundaryArea WRITE setBoundaryArea)
+	PYB11_PROPERTY(float PrecipBoundaryFraction READ getPrecipBoundaryFraction WRITE setPrecipBoundaryFraction)
+	PYB11_METHOD(QString getStatsType)
```
Indentation

mmarineBlueQuartz commented on Source/SIMPLib/StatsData/PrimaryStatsData.h:
```diff
@@ -82,6 +82,19 @@ typedef QVector<FloatArrayType::Pointer> VectorOfFloatArray;
+	PYB11_METHOD(PhaseType::Type getPhaseType)
+	PYB11_METHOD(FloatArrayType::Pointer generateBinNumbers)
+	PYB11_PROPERTY(float PhaseFraction READ getPhaseFraction WRITE setPhaseFraction)
+	PYB11_PROPERTY(float BoundaryArea READ getBoundaryArea WRITE setBoundaryArea)
```
Indentation

mmarineBlueQuartz commented on Source/SIMPLib/StatsData/StatsData.h:
```diff
@@ -89,6 +89,11 @@ typedef QVector<FloatArrayType::Pointer> VectorOfFloatArray;
 {
+	PYB11_CREATE_BINDINGS(StatsData)
+	PYB11_METHOD(QString getStatsType)
+	PYB11_METHOD(PhaseType::Type getPhaseType)
```
Indentation

mmarineBlueQuartz commented on Wrapping/Python/Pybind11/CodeScraper/PythonBindingClass.cpp:
```diff
@@ -432,7 +432,11 @@ QString PythonBindingClass::generateStaticNewCode()
+	  if (tokens.size() < 2 || tokens[1] != ::kOverload)
+	  {
+		  constructors << TAB << ".def_static(\"" << methodName << "\", &" << getClassName() << "::" << methodName << ")" << NEWLINE_SIMPL;
+	  }
```
Non-standard indentation.  Are tabs being used here?

mmarineBlueQuartz commented on Wrapping/Python/Pybind11/CodeScraper/PythonBindingClass.cpp:
```diff
@@ -562,8 +566,17 @@ QString PythonBindingClass::generatePropertiesCode()
+  if (methodArgs.contains("lambda") || bodyCodes.contains("lambda"))
+  {
+	  methodArgs.replace("lambda", "lambda_1");
+	  bodyCodes.replace("lambda", "lambda_1");
```
Indentation

mmarineBlueQuartz commented on Wrapping/Python/Pybind11/CodeScraper/PythonBindingClass.cpp:
```diff
@@ -620,16 +633,26 @@ QString PythonBindingClass::generateMethodCode()
+		returnValuePolicy = ", " + tokens[index + 1];
+		tokens.removeAt(index);
+		tokens.removeAt(index);
+	}
```
Indentation

mmarineBlueQuartz commented on Wrapping/Python/Pybind11/CodeScraper/PythonBindingsModule.cpp:
```diff
@@ -164,6 +166,7 @@ void PythonBindingsModule::addDependency(QString superClassName, QString classNa
   {
     QObject* obj = new QObject(nullptr);
     obj->setObjectName(className);
+	obj->setProperty("superClass", superClassName);
```
Indentation.  Lines 155, 160, and 169

mmarineBlueQuartz commented on Wrapping/Python/Pybind11/CodeScraper/PythonBindingsModule.cpp:
```diff
@@ -359,8 +358,11 @@ void PythonBindingsModule::generatePythonTestFile(const QString& outputPath, con
+	  {
+		  out << "\n";
+		  dumpRecursivePythonCode(0, object, out);
+	  }
```
Indentation

mmarineBlueQuartz commented on Wrapping/Python/Pybind11/CodeScraper/PythonBindingsModule.cpp:
```diff
@@ -417,9 +422,12 @@ void PythonBindingsModule::generatePythonicInterface(const QString& outputPath,
     }
+	else if (bodyCodes.isEmpty()) {
+		continue;
+	}
```
Indentation

mmarineBlueQuartz commented on Wrapping/Python/Pybind11/Templates/Pybind11CustomTypeCasts.in.h:
```diff
@@ -0,0 +1,620 @@
+		};
+
+	} // namespace detail
+} // namespace pybind11
```
Not sure if we care about .in files, but I think you're using tab

mmarineBlueQuartz reviewed (changes requested):
There are definitely a lot of places indentation gets thrown off.  Some of it is indentation from macros, and some of it looks like you are using tabs instead of spaces.

A bigger concern of mine is the non-standard lowercase method names like `getdataContainers`.  I started to think that was intentional, but then I found more PyBind methods that were added that followed our standard naming (i.e. along the lines of `getDataContainers`).  I am now confused on what's what, why we're doing it that way sometimes and not others, and if there's a specific reason, can we please mark it each time in the `@brief` and in the .cpp so no one breaks the python binding by "fixing" it?  Because that's probably going to be me.  That, and being inconsistent in our Python API is probably not a good thing either.

I did not review the Python files.  I know I am not the right person for that.

Also, the three line comment marker we put before methods in .cpp files is missing from many of the new methods but not all.

imikejackson commented:
the lowercase method names were updated appropriately. Those were non-conforming and not caught during development.

The indentations are getting taken care of. This might be a good time to just run clang format over the code base.

imikejackson committed:
Code formatting and proper camelCase implemented.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fixing more capitalization and indentation

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

mmarineBlueQuartz commented on Source/SIMPLib/DataArrays/DataArray.hpp:
```diff
@@ -855,6 +855,31 @@ class DataArray : public IDataArray
+		int i = 0;
+		for (auto elem : newArray)
+		{
+			m_Array[i++] = elem;
```
Clang format getArray and setArray

imikejackson replied:
I'm going to clang-format the entire file.

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/AttributeMatrixProxy.cpp:
```diff
@@ -285,3 +278,63 @@ void AttributeMatrixProxy::updatePath(DataArrayPath::RenameType renamePath)
+
+QMap<QString, DataArrayProxy> AttributeMatrixProxy::getDataArrays()
+{
+	return dataArrays;
```
getDataArrays through setFlag all have extra indentation and are missing the comment block above their definitions unlike other methods.

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/DataArrayProxy.cpp:
```diff
@@ -324,3 +324,23 @@ void DataArrayProxy::updatePath(DataArrayPath::RenameType renamePath)
+
+void DataArrayProxy::setFlag(uint8_t newFlag)
+{
+	flag = newFlag;
```
Indentation is still off for getFlag and setFlag.  Comment blocks are missing above method definitions between getFlag and getName

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/DataContainer.h:
```diff
@@ -73,8 +73,8 @@ class SIMPLib_EXPORT DataContainer : public Observable
-  PYB11_METHOD(AttributeMatrix::Pointer getAttributeMatrix OVERLOAD const.QString.&, Name)
-  PYB11_METHOD(AttributeMatrix::Pointer getAttributeMatrix OVERLOAD const.DataArrayPath.&, Path)
+  PYB11_METHOD(AttributeMatrix::Pointer getAttributeMatrix OVERLOAD const.QString.&,Name)
+  PYB11_METHOD(AttributeMatrix::Pointer getAttributeMatrix OVERLOAD const.DataArrayPath.&,Path)
```
Is the space after the comma important at the end of this macro?

imikejackson replied:
YES. The space MUST be there for our parsing to work.

mmarineBlueQuartz replied:
The space is missing again.

imikejackson replied:
Actually, no the space should NOT be there. I am working a better solution to this.

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/DataContainerArrayProxy.cpp:
```diff
@@ -544,3 +544,13 @@ QMap<QString, DataContainerProxy> DataContainerArrayProxy::readMap(QJsonArray js
+
+QMap<QString, DataContainerProxy> DataContainerArrayProxy::getDataContainers()
+{
+	return dataContainers;
```
getDataContainers and setDataContainers still require indentation checks.  The comment blocks above their definitions are also missing.

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/DataContainerProxy.cpp:
```diff
@@ -228,10 +228,81 @@ void DataContainerProxy::setFlags(uint8_t flag, AttributeMatrixProxy::AMTypeFlag
+	for (QMap<QString, AttributeMatrixProxy>::iterator amIter = attributeMatricies.begin(); amIter != attributeMatricies.end(); ++amIter) // AttributeMatrix Level
+	{
+		AttributeMatrixProxy& am = amIter.value();
+    if(am.getName() == name)
```
An if statement inside a for loop should not be less indented than the for that contains it.

mmarineBlueQuartz replied:
Also, `AttributeMatrixProxy proxy(name);` is oddly indented a few lines after this to match the for loop.

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/DataContainerProxy.h:
```diff
@@ -35,141 +35,203 @@
+  uint8_t flag;
+  QString name;
+  unsigned int dcType;
+  QMap<QString, AttributeMatrixProxy> attributeMatricies;
```
We have getters and setters for all of these variables, and SIMPL code has been updated to use them.  Do we still want these to be public variables, or private?

imikejackson replied:
Some of the codes still directly access them and I have not had the time to work through the ramifications of making them private (which I agree that we should do).

imikejackson committed:
Additional code cleanup

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

mmarineBlueQuartz reviewed (approved):
This looks good to me.

imikejackson committed:
Refactored the public variables in the *Proxy classes

+ All the public variables that were available publicly are now private and accessed
with setters/getters. Added API to 'insert' an object into the storage class in
each of the proxy classes.

+ Added comments to disable clang-formatting for the PYBIND11 macro areas as we depend
on specific formatting around the commas.
Adding in the clang on|off switches forces clang format to reformat certain files.

+ Most loops in the proxy classes were changed to range-based C++11 loops.

+ FilterParameters that use a struct class to hold data, the struct declaration and
definition was moved to its own header file. The python bindings code (pybind11)
was cleaned up due to this.

+ Additional cleanups in the template pybind11 codes to remove dependencies on OrientationLib

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/AttributeMatrix.h:
```diff
@@ -96,7 +97,10 @@ class SIMPLib_EXPORT AttributeMatrix : public Observable
   PYB11_METHOD(int renameAttributeArray ARGS OldName NewName OverWrite)
-  PYB11_METHOD(IDataArray::Pointer getAttributeArray ARGS const.QString.&,name RETURN_VALUE_POLICY py::return_value_policy::reference)
+  PYB11_METHOD(IDataArray::Pointer getAttributeArray OVERLOAD const.QString.&,Name)
+  PYB11_METHOD(IDataArray::Pointer getAttributeArray OVERLOAD const.DataArrayPath.&,Path)
```
Missing a space between the comma and Name / Path

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/DataArrayProxy.h:
```diff
@@ -171,28 +180,86 @@ class SIMPLib_EXPORT DataArrayProxy
    * @param name
    */
-  void setName(const QString& name);
+  void setName(const QString& m_Name);
```
Are you sure about this rename?

imikejackson replied:
Fixed

mmarineBlueQuartz commented on Source/SIMPLib/DataContainers/DataContainerProxy.h:
```diff
@@ -193,7 +201,7 @@ class SIMPLib_EXPORT DataContainerProxy
    * @return the attribute matrix proxy
    */
-  AttributeMatrixProxy& getAttributeMatrixProxy(const QString& name);
+  AttributeMatrixProxy& getAttributeMatrixProxy(const QString& m_Name);
```
Are you sure about renaming the parameters in setFlag and getAttributeMatrixProxy?  If not, you might want to update the insertAttributeMatrix declaration as well.

mmarineBlueQuartz replied:
Same with the flag parameter in setFlags below.

mmarineBlueQuartz commented on Source/SIMPLib/Filtering/ComparisonSet.h:
```diff
@@ -53,6 +54,7 @@ class SIMPLib_EXPORT ComparisonSet : public AbstractComparison
   PYB11_PROPERTY(int UnionOperator READ getUnionOperator WRITE setUnionOperator)
   PYB11_METHOD(void addComparison ARGS AbstractComparisonPtr,comparison)
   PYB11_METHOD(void insertComparison ARGS int,index AbstractComparisonPtr,comparison)
```
I am going to assume spaces are important here, too.  Did I miss this in the earlier commits, or is not having a space actually valid here?

mmarineBlueQuartz commented on Source/SIMPLib/Filtering/ComparisonSet.h:
```diff
@@ -53,6 +54,7 @@ class SIMPLib_EXPORT ComparisonSet : public AbstractComparison
   PYB11_PROPERTY(int UnionOperator READ getUnionOperator WRITE setUnionOperator)
   PYB11_METHOD(void addComparison ARGS AbstractComparisonPtr,comparison)
   PYB11_METHOD(void insertComparison ARGS int,index AbstractComparisonPtr,comparison)
+  // clang-format om
```
`// clang-format on` not om

mmarineBlueQuartz commented on Source/SIMPLib/DataArrays/StringDataArray.h:
```diff
@@ -63,6 +64,8 @@ class SIMPLib_EXPORT StringDataArray : public IDataArray
   PYB11_METHOD(void setValue ARGS size_t,i const.QString.&,value)
```
Are additional spaces necessary here?

imikejackson replied:
Yes

mmarineBlueQuartz commented on Source/SIMPLib/Filtering/ComparisonInputs.h:
```diff
@@ -97,18 +97,16 @@ typedef struct
   PYB11_PROPERTY(QVector<ComparisonInput_t> Inputs READ getInputs)
-
   PYB11_METHOD(void addInput OVERLOAD const.QString,dataContainerName const.QString,attributeMatrixName const.QString,arrayName int,compOperator double,compValue)
   PYB11_METHOD(void addInput OVERLOAD const.ComparisonInput_t.&,input)
```
I believe you need a space before `input`.

imikejackson replied:
Nope.

mmarineBlueQuartz reviewed (changes requested):
Minimum Requirement:

- Fix one `// clang-format om`

Strongly Suggested:

- Put const back in methods that removed it from proxy parameters.  Within those methods, please use them for each range-based for loop.  Obviously, not of these cases make sense to be const, but the amount of non-const references in places that should not change the value makes me worry about how easy it would be to break pipelines or UI.
   - Add const versions (const T& name() const;)  of each proxy getter that returns a reference so that const can be used where appropriate.  For instance, adding `const QMap<QString, DataArrayProxy>& getDataArrays() const;` to AttributeMatrixProxy in addition to the non-const version.  The lack of having these methods declared in both versions seems to be why const was removed from proxy parameters.
   - Logic behind this: If something should not be changed, make it a compiler error to do so.  This is why each of the STL containers have const and non-const versions defined for most getters.
   - If you do not want to duplicate code between const and non-const methods, an acceptable solution is to call the const version from non-const method and cast away const on the return statement.  It is not pretty, but it is safe because the object itself (and thus its variables) are not promised to be constant when calling a non-const method.

Edit: Apparently space requirements in PYB11_METHOD have changed.

imikejackson commented:
In some places we do need to edit the underlying storage. Maybe we should then just return an iterator that allows outside codes to modify the underlying storage? That was the idea behind add the "insert*()" methods into the proxy classes.

mmarineBlueQuartz commented:
Iterators would be a good idea as long as we have both iterator and const_iterator versions (QMap already offers both). It would be a more convenient way to simply loop over a proxy without having to create additional variables.

imikejackson committed:
Changes to the Proxy classes to add a const method to get the StorageType.

Also updates to the pybind11 generation codes.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Update to the Anaconda creation workflow.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Adjusting output folders for the build Anaconda package.

imikejackson committed:
Additional changes to how the conda package is built.

Using a shell script on macOS/Linux and will eventually use a batch file under
Windows to drive the creation of the conda package.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Edits to the generated batch files to ensure that they will work

Batch files are finicky.

imikejackson committed:
More updates to the macOS conda packaging scripts.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
More refinements to the macOS packaging script file.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>
